File: PyAlgorithms/recursive/fibonacci.py
# https://leetcode-cn.com/problems/fibonacci-number/
class Solution(object):
    # 递归，时间复杂度为O(n^2)
    def fib1(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n == 0 or n == 1:
            return n

        return self.fib1(n - 1) + self.fib1(n - 2)

    d = {}

    # 递归，但使用哈希表缓存已经计算过的值，时间复杂度为O(n)
    def fib(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n == 0 or n == 1:
            return n

        if n in self.d:
            return self.d[n]
        else:
            res = self.fib(n - 1) + self.fib(n - 2)
            self.d[n] = res
            return res


if __name__ == '__main__':
    s = Solution()
    print(s.fib(4))

File: CppAlgorithms/recursive/Fibonacci.cpp
#include <unordered_map>
#include <iostream>
using namespace std;
/**
 * https://leetcode-cn.com/problems/fibonacci-number/
 *
 * 斐波那契数，通常用 F(n) 表示，形成的序列称为 斐波那契数列
 * 该数列由 0 和 1 开始，后面的每一项数字都是前面两项数字的和。也就是：
 * F(0) = 0，F(1) = 1
 * F(n) = F(n - 1) + F(n - 2)，其中 n > 1
 *
 * 给一个 n ，请计算 F(n)
 */
class Fibonacci {
private:
    unordered_map <int, int> map;
public:
    int fib(int n) {
        if (n == 0 | n == 1)
            return n;
        if (map.find(n) == map.end()){
            int res = fib(n-1) + fib(n-2);
            pair<int, int> kv(n, res);
            map.insert(kv);
            return res;
        } else{
            return map[n];
        }
    }
};

int main(){
    Fibonacci s;
    int res = s.fib(6);
    std::cout << res << std::endl;
}